File: generator/src/main/java/ch/bailu/gtk/model/ParameterModel.java
package ch.bailu.gtk.model;

import static ch.bailu.gtk.converter.UtilKt.isEnum;
import static ch.bailu.gtk.log.UtilKt.colonList;
import static ch.bailu.gtk.model.filter.FilterKt.values;

import ch.bailu.gtk.converter.JavaNames;
import ch.bailu.gtk.converter.JniTypeConverter;
import ch.bailu.gtk.model.type.CType;
import ch.bailu.gtk.model.type.ClassType;
import ch.bailu.gtk.model.type.JavaType;
import ch.bailu.gtk.tag.ParameterTag;

public class ParameterModel extends Model {
    private final String name;

    private final CType cType;
    private final ClassType classType;
    private final JavaType jType;

    //private final String value;

    private final JniTypeConverter jniConverter;

    final private boolean isWriteable;

    private MethodModel callbackModel = null;

    private final ParameterTag parameterTag;

    public ParameterModel(String namespace, ParameterTag parameterTag, boolean toUpper, boolean supportsDirectAccess) {

        this.parameterTag = parameterTag;

        if (toUpper) {
            name = JavaNames.fixToken(parameterTag.getName().toUpperCase());
         } else {
            name = JavaNames.fixToken(parameterTag.getName());
        }

        classType = new ClassType(namespace, parameterTag, supportsDirectAccess);
        if (classType.isClass()) {
            cType = new CType("void*");
            jType = new JavaType("long");

        } else if (isEnum(namespace, parameterTag)) {
            cType = new CType("int");
            jType = new JavaType("int");

        } else {
            cType = new CType(parameterTag.getType());
            jType = new JavaType(parameterTag.getType());
        }

        if (classType.isCallback()) {
            callbackModel = new MethodModel(namespace, classType.getCallbackTag());
        }


        jniConverter = JniTypeConverter.factory(this);

        //setSupported("private", parameter.isPrivate());
        setSupported("value", values(getValue()));
        setSupported("jType", jType.isValid());
        setSupported("callback", isCallbackSupported());

        this.isWriteable = parameterTag.isWriteable();
    }


    private boolean isCallbackSupported() {
        if (isCallback()) {
            return callbackModel.isSupported() && !callbackModel.hasCallback();
        }
        return true;
    }

    public String getValue() {
        return saveString(parameterTag.getValue());
    }
    public String getName() {
        return saveString(name);
    }

    public String getApiType() {
        if (isCallback()) {
            return callbackModel.getSignalInterfaceName();
        }

        if (classType.isClass()) {
            return classType.getFullName();
        }

        return jType.getType();
    }

    public String getImpType() {
        return jType.getType();
    }

    public boolean isVoid() {
        return jType.isVoid();
    }

    public boolean isJavaNative() {
        return classType.isClass() == false;
    }

    public String getJniType() {
        return jniConverter.getJniType();
    }



    public String getGtkType() {
        return cType.getType();
    }


    @Override
    public String toString() {
        String supported="";
        if (isSupported()) {
            supported = "(s)";
        }

        return colonList(new String[] {
                supported,
                parameterTag.getType(),
                parameterTag.getTypeName(), // <-
                parameterTag.getName(),     // <-
                parameterTag.getValue(),
                getGtkType(),
                getApiType()});
    }


    private static String saveString(String in) {
        if (in == null) return "";
        return in;
    }

    public String getImpDefaultConstant() {
       return jniConverter.getImpDefaultConstant();
    }

    public String getJniSignatureID() {
        return jniConverter.getJniSignatureID();
    }


    public String getJniCallbackMethodName() {
        return jniConverter.getJniCallbackMethodName();
    }

    public boolean isWriteable() {
        return isWriteable;
    }

    public boolean isCallback() {
        return classType.isCallback();
    }

    public MethodModel getCallbackModel() {
        return callbackModel;
    }

    public JniTypeConverter getJniConverter() {
        return jniConverter;
    }

    public boolean isDirectType() {
        return classType.isDirectType();
    }
}

File: generator/src/main/kotlin/ch/bailu/gtk/model/ParameterModel.kt
package ch.bailu.gtk.model

import ch.bailu.gtk.log.DebugPrint
import ch.bailu.gtk.model.filter.filterValues
import ch.bailu.gtk.model.type.CType
import ch.bailu.gtk.model.type.CallbackType
import ch.bailu.gtk.model.type.ClassType
import ch.bailu.gtk.model.type.JavaType
import ch.bailu.gtk.validator.Validator
import ch.bailu.gtk.parser.tag.ParameterTag
import ch.bailu.gtk.table.EnumTable
import ch.bailu.gtk.writer.Names

class ParameterModel(namespace: String,
                     private val parameterTag: ParameterTag,
                     isConstant: Boolean,
                     preferNative: Boolean) : Model() {

    private val classType = ClassType(namespace, parameterTag)
    private val callbackType = CallbackType(namespace, parameterTag.getTypeName())
    private val jType: JavaType
    val hasNativeVariant: Boolean
    val isNativeVariant: Boolean

    var name: String = if (isConstant) {
        Names.getJavaConstantName(parameterTag.getName())
    } else {
        Names.getJavaVariableName(parameterTag.getName())
    }

    val callbackModel: MethodModel?

    init {
        if (!classType.valid && EnumTable.isEnum(namespace, parameterTag)) {
            this.jType = JavaType("int")
            hasNativeVariant = false

        } else {
            val jType = JavaType(parameterTag.getType())
            hasNativeVariant = classType.valid && jType.valid

            if (hasNativeVariant && preferNative) {
                this.jType = jType

            } else if (classType.valid) {
                this.jType = JavaType("long")

            } else if (parameterTag.isVarargs) {
                this.jType = JavaType("...")
            } else {
                this.jType = jType
            }
        }

        isNativeVariant = hasNativeVariant && preferNative

        callbackModel = createCallbackModel(classType, namespace)

        setSupported("value-not-supported", filterValues(parameterTag.value))
        setSupported("java-type-not-supported", jType.valid || isCallback)
        setSupported("direct-type", classType.referenceType || classType.wrapper || isCallback)
        setCallbackSupported()
    }

    private fun createCallbackModel(classType: ClassType, namespace: String): MethodModel? {
        val parameterNamespace = classType.type.namespace.ifEmpty {
            namespace
        }

        if (callbackType.valid) {
            if (callbackType.callbackTag != null) {
                return MethodModel(namespace, parameterNamespace, callbackType.callbackTag, preferNative = false)
            }
        }
        return null
    }

    private fun setCallbackSupported() {
        if (isCallback) {
            if (callbackModel == null) {
                Validator.giveUp("no-cb-model")
            } else if (!callbackModel.isSupported) {
                setSupported("cb-not-supported", false)
            } else if (callbackModel.hasCallback()) {
                setSupported("cb-with-cb", false)
            }
        }
    }

    /**
     * Return valid and conventional Java type name or native type name
     * Add namespace to classes if they are in a different namespace
     * @param namespace namespace where other namespace is relative for
     */
    fun getApiTypeName(namespace: String): String {
        return if (isCallback && callbackModel != null) {
            Names.getJavaCallbackInterfaceName(callbackModel.name)
        } else if (classType.valid && !isNativeVariant) {
            classType.getApiTypeName(namespace)
        } else {
            jType.getApiTypeName()
        }
    }

    val impType: String
        get() {
            return jType.getApiTypeName()
        }

    val isVoid: Boolean
        get() = jType.isVoid()

    val isJavaNative: Boolean
        get() {
            return !classType.valid || isNativeVariant
        }


    override fun toString(): String {
        return if (isCallback) {
            DebugPrint.colon(this, supportedState, callbackModel.toString())

        } else {
            DebugPrint.colon(this,
                supportedState,
                classType.toString(),
                jType.toString(),
            )
        }
    }

    val isCallback: Boolean
        get() = callbackType.valid

    val value: String
        get() = parameterTag.value

    val doc: String
        get() = parameterTag.getDoc()

    val nullable: Boolean
        get() = parameterTag.nullable
}
